SDWebImage/SDWebImageManager.m

```diff
@@ -292,27 +309,27 @@ - (void)saveImageToCache:(nullable UIImage *)image forURL:(nullable NSURL *)url
 }
 
 - (void)cancelAll {
-    @synchronized (self.runningOperations) {
```

Comment:
This one can use semaphore as well. And it's not actually need a recursive lock :)

Because all you need is to get the array of running operations, you don't have to lock all the process including `cancel`.

So the code can be like this:

```objective-c
LOCK(self.runningOperationsLock);
NSArray<SDWebImageCombinedOperation *> *copiedOperations = [self.runningOperations copy];
UNLOCK(self.runningOperationsLock);
[copiedOperations makeObjectsPerformSelector:@selector(cancel)]; // This will call `safelyRemoveOperationFromRunning`
```

Reply:
Code updated, please check out again :)